File: typescript/CDO/Resource.ts
var xmpp = require('../node-xmpp/packages/client').xmpp;
var xml = require('../node-xmpp/packages/xml');
var Guid = require('../Misc/Guid');
import {CDO} from './CDO';
import {DataVerb} from './DataVerb';

export class Resource 
{
    public Path:string;
    public Capabilities : Array<string> = new  Array<string>();
    public MetaAttributes : {[key: string] : string} =  {}; 
    public SupportedVerbs: Array<DataVerb> = new Array<DataVerb>();

    
    public GetXML() : any
    {
        let RVal = xml("resource", {xmlns:CDO.Namespace});

        CDO.AddResourcePath(RVal, "path", this.Path);
        let CapEL = CDO.StartList(RVal, "capabilities");
        for (let Cap of this.Capabilities)
        {
            CDO.SetString(CapEL, Cap);
        }
        if( Object.keys(this.MetaAttributes).length > 0)
        {
            let DictEl = CDO.StartDictionary(RVal, "metaattributes");
            for(let item in this.MetaAttributes)
            {
                let ItemEl = CDO.StartDictionaryItem(DictEl);
                CDO.AddString(ItemEl, "key", item);
                CDO.AddString(ItemEl, "value", this.MetaAttributes[item]);
            }
        }
        let SupVerbEl = CDO.StartList(RVal, "supportedverbs");
        for(let verb of this.SupportedVerbs)
        {
            CDO.SetDataVerb(SupVerbEl, verb);
        }

        return RVal;
    }
}

File: NET/CDO/Resource.cs
using System;
using System.Collections.Generic;
using System.Xml.Linq;

namespace CDO
{
	public class Resource
	{
		public string Path;
		public List<string> Capabilities = new List<string>();
		public Dictionary<string, string> MetaAttributes = new Dictionary<string, string>();
		public List<DataVerb> SupportedVerbs = new List<DataVerb>();

		public XElement GetXML()
		{
			XElement RVal = new XElement(CDO.Namespace + "resource");

			CDO.SetResourcePath(RVal, "path", Path);
			XElement CapEL = CDO.StartList(RVal, "capabilities");
			foreach (string Cap in Capabilities)
			{
				CDO.SetString(CapEL, Cap);
			}

			if (MetaAttributes != null && MetaAttributes.Count > 0)
			{
				XElement DictEl = CDO.StartDictionary(RVal, "metaattributes");
				foreach (var item in MetaAttributes)
				{
					XElement ItemEl = CDO.StartDictionaryItem(DictEl);
					CDO.SetString(ItemEl, "key", item.Key);
					CDO.SetString(ItemEl, "value", item.Value);
				}
			}

			XElement SupVerbEl = CDO.StartList(RVal, "supportedverbs");
			foreach (DataVerb verb in SupportedVerbs)
			{
				CDO.SetDataVerb(SupVerbEl, verb);
			}

			return RVal;
		}
	}
}
